tao_compiler/mlir/disc/transforms/disc_transpose_simplifier.cc

```diff
@@ -882,6 +922,141 @@ LogicalResult reverseBinaryOpsIfBeneficial(Block* block, bool& changed) {
+// and apply the rewrite when it's beneficial.
+LogicalResult tryMoveUpTransposeGreedilyAndApplyIfBeneficial(Block* block,
+                                                             bool& changed) {
+  TransposeSimpliferContext ctx(block);
+  for (auto op : ctx.opList) {
+    auto transposeOp = dyn_cast<mhlo::TransposeOp>(op);
+    if (!transposeOp) continue;
+    LLVM_DEBUG(llvm::dbgs()
+               << "tryMoveUpTransposeGreedily for " << *op << "\n");
+
+    int numTransposeEliminated = 0;
+    DenseMap<Operation*, SmallVector<int64_t>> permMap;
+    DenseMap<Operation*, DenseSet<Operation*>> userMap;
+    std::priority_queue<std::pair<int, Operation*>> queue;
+    auto tryUpdate = [&](Operation* producer, Operation* consumer,
+                         const SmallVector<int64_t>& perm) {
+      // do not propagate transpose across block a.t.m.
+      if (!ctx.inTargetBlock(producer)) return true;
+      auto it = permMap.find(producer);
+      if (it != permMap.end()) {
+        if (!isa<mhlo::ConstantOp>(producer) && it->second != perm)
+          return false;
+        userMap[producer].insert(consumer);
+        return false;
+      } else {
+        permMap[producer] = perm;
+        userMap[producer].insert(consumer);
+        queue.emplace(
+            std::pair<int, Operation*>{ctx.op2Idx[producer], producer});
+        return true;
+      }
+    };
+    tryUpdate(
+        op->getOperand(0).getDefiningOp(), op,
+        llvm::to_vector<>(transposeOp.getPermutation().getValues<int64_t>()));
+
+    while (!queue.empty()) {
+      Operation* curOp = queue.top().second;
+      queue.pop();
+      LLVM_DEBUG(llvm::dbgs() << "\tcurOp: " << *curOp << "\n");
+      LLVM_DEBUG(llvm::dbgs() << "\tinit numTransposeEliminated for curOp: "
+                              << numTransposeEliminated << "\n");
+      // TODO(kevin.zwy): support op with more than one result.
+      if (curOp->getNumResults() > 1) continue;
+      // check if we can move up transpose op through this op.
+      DenseMap<int, SmallVector<int64_t>> operandPermMap;
+      if (failed(backwardPermutation(curOp, permMap[curOp], operandPermMap)))
+        continue;
+      LLVM_DEBUG(llvm::dbgs() << "\tpass backwardPermutation for curOp\n");
+
+      bool needReverse = false;
+      auto& transposedUsers = userMap[curOp];
+      LLVM_DEBUG(llvm::dbgs() << "\ttransposedUsers list of curOp ("
+                              << transposedUsers.size() << "):\n");
+      for (auto& transposedUser : transposedUsers) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "\t\t transposedUser: " << *transposedUser << "\n");
+      }
+      for (auto user : curOp->getResult(0).getUsers()) {
+        LLVM_DEBUG(llvm::dbgs() << "\tuser of curOp: " << *user << "\n");
+        if (isa<tensor::DimOp, shape::ShapeOfOp, mhlo::TransposeOp>(user))
+          continue;
+        if (transposedUsers.find(user) != transposedUsers.end()) continue;
+        needReverse = true;
```

Comment:
nit: break?

Reply:
It's the last statement of the for loop body, thus I think it's not necessary to add a break here.